### PR TITLE
Fix check by delivery price/weight in case of hook usage

### LINKED ISSUE #27158: Wrong usage of `actionDeliveryPriceByPrice` and `actionDeliveryPriceByWeight` in Carrier class
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add screenshots

In Carrier class `checkDeliveryPriceByWeight` and `checkDeliveryPriceByPrice` should return a boolean.
`actionDeliveryPriceByPrice` or `actionDeliveryPriceByWeight` hooks returns shipping price.

If this hooks return 0 (free delivery) `checkDeliveryPriceByWeight` and `checkDeliveryPriceByPrice` return `false` (0) instead of `true`.
So carriers disapear front customer choice.

### Expected behavior

return true of false

### Steps to reproduce

1. Hook a module to actionDeliveryPriceByPrice 
2. Retun 0
3. See the carrier disapear ( the carrier has to be set like this : Out-of-range behavior : Disable carrier)

image

### PrestaShop version(s) where the bug happened

1.7.x

### PHP version(s) where the bug happened

all

### If your bug is related to a module, specify its name and its version

_No response_

### PR DESCRIPTION
**Pull request by jf-viguier**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | In Carrier class, checkDeliveryPriceByWeight and checkDeliveryPriceByPrice should return a boolean. actionDeliveryPriceByPrice or actionDeliveryPriceByWeight hooks returns shipping price that are simply returned causing a bug if shipping is free.
| Type?             | bug fix
| Category?         | FO
| BC breaks?        | yes
| Deprecations?     | no
| Fixed ticket?     | Fixes https://github.com/PrestaShop/PrestaShop/issues/27158
| How to test?      | See issue
| Possible impacts? | none

BC :
checkDeliveryPriceByWeight() and checkDeliveryPriceByPrice() now return only boolean. Use getDeliveryPriceByWeight() and getDeliveryPriceByPrice() to get the values.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/27159)

### CODE REVIEW

File: classes/Carrier.php [hunk 1]
```diff
@@ -327,7 +327,7 @@ public function getDeliveryPriceByWeight($total_weight, $id_zone)
      * @param float $total_weight Total weight
      * @param int $id_zone Zone ID
      *
-     * @return float|bool Delivery price, false if not possible
+     * @return bool true if carrier is available
```
Comment (PierreRambaud) [inline classes/Carrier.php:330]: It's a BC break as the value is now a boolean :thinking:

**Review by matthieu-rolland — COMMENTED** [review]
Thank you @jf-viguier for your fix !

The idea behind your fix is good, but as @PierreRambaud said, it's a BC

As I tried to understand the goal of this method, I checked in our codebase how it is used, and the result of this method is always used as a boolean indeed 😓 So I understand where your fix comes from

Still, keeping the name of this function as is and returning a boolean makes it very ambiguous 🤔 

Maybe it's better to add a new method that returns a boolean, like "isFreeDeliveryForGivenWeight", and call `checkDeliveryPriceByWeight` inside, returning a boolean. 

Because that's the information we want in the end, is there a delivery cost or not...

What do you think ?

**Comment by jf-viguier** [comment]
The 2 methods always return boolean : the php doc is wrong.
The hook is called twice : in the check and in the get, so it's buggy.
If the hook is called : it return the shipping cost, otherwise true or false (isset).

Is a buggy phpdoc is still a BC ? I don't know

File: classes/Carrier.php (hunk 1 above)
Comment (Progi1984) [reply classes/Carrier.php:330]: I disagree as the signature doesn't change.

Comment (matthieu-rolland) [reply classes/Carrier.php:330]: sometimes the method returned zero `0`, now it will return `false` instead, doesn't that make it a breaking change ?

If it's a breaking change, it's minor anyway, since the result is used as a boolean in our codebase anyway. But it should still be noted in the PR description.

**Review by matthieu-rolland — DISMISSED** [review]
thanks for fixing the phpdoc @jf-viguier , it's indeed much more coherent like this 👍

File: classes/Carrier.php (hunk 1 above)
Comment (jf-viguier) [reply classes/Carrier.php:330]: @matthieu-rolland I've documented this

**Review by matks — DISMISSED** [review]
Hi @jf-viguier thanks for this contribution

I'm sorry but I do not understand why you write `In Carrier class, checkDeliveryPriceByWeight and checkDeliveryPriceByPrice should return a boolean.`

Why should we return a boolean? Why is this an issue if sometimes a numeric value is returned?

> If this hooks return 0 (free delivery) checkDeliveryPriceByWeight and checkDeliveryPriceByPrice return false (0) instead of true.

This can be solved by simply casting the result when the shipping price is 0.

It would really be helpful if you could provide the context for this PR 😉  see [this example](https://twitter.com/lorismatic/status/1469308514170556416?t=iuaOwLJk6FXfkI-HSMDBfA&s=19)

**Comment by jf-viguier** [comment]
> I'm sorry but I do not understand why you write In Carrier class, checkDeliveryPriceByWeight and checkDeliveryPriceByPrice should return a boolean.
> Why should we return a boolean? Why is this an issue if sometimes a numeric value is returned?

@matks have you check the issue ? 

checkDeliveryPriceByWeight() and checkDeliveryPriceByPrice() methods are used in Carrier and Cart classes this way:
```
if (Carrier::checkDeliveryPriceByPrice(..) ...) {
    unset($carrier);
```
here by example : https://github.com/PrestaShop/PrestaShop/blob/e9f35494ccc232fcba6f92ec1605bdf3d5481ce5/classes/Carrier.php#L757
    
    
So, if the hook return 0 (free shipping) the carrier is skipped and disapears in front. It should be displayed and his price set to 0€.

**Comment by matks** [comment]
@jf-viguier Today, this method can return 3 values
- a boolean
- integer zero
- an integer different from zero

You modify the behavior to handle the usecase "integer zero". But you also remove the 3rd usecase, when a not-zero integer is returned. This is why I dont understand.

**Comment by jf-viguier** [comment]
@matks please don't focus on method phpdoc, it's wrong.
Go to where the methods are used and see by yourself.

My undestanding on how it should be:
checkDeliveryPriceByWeight() : is this carrier is available or not ? Should return true or false
getDeliveryPriceByPrice() : return shipping price 

Actual behaviour :
checkDeliveryPriceByWeight() without module on the hook => return true or false
checkDeliveryPriceByWeight() with module on the hook => return the shipping price

getDeliveryPriceByPrice() with module on the hook => return shipping price 

In other words : a module can't manage free shipping: this will cause the carrier to disappears.

Check the video here, bug is reproduced now https://github.com/PrestaShop/PrestaShop/issues/27158#issuecomment-1006796148

**Comment by jf-viguier** [comment]
@matthieu-rolland said :
> As I tried to understand the goal of this method, I checked in our codebase how it is used, and the result of this method is always used as a boolean indeed 😓 So I understand where your fix comes from

**Comment by matks** [comment]
> please don't focus on method phpdoc, it's wrong.
Go to where the methods are used and see by yourself.

I dont read the phpdoc. I compare behavior in `develop` and behavior in your contribution.

If this method returns the integer zero, and zero is badly casted into `false` by a wrong comparison, the comparison must be fixed instead.

> If this hooks return 0 (free delivery) checkDeliveryPriceByWeight and checkDeliveryPriceByPrice return false (0) instead of true.

False and zero are not the same thing.

Let's see one usage of `checkDeliveryPriceByPrice()`

```
// Carrier.php line 757
                    if ($shipping_method == Carrier::SHIPPING_METHOD_PRICE
                        && (!Carrier::checkDeliveryPriceByPrice($row['id_carrier'], $cart->getOrderTotal(true, Cart::BOTH_WITHOUT_SHIPPING), $id_zone, $id_currency ?? null))) {
                        $error[$carrier->id] = Carrier::SHIPPING_PRICE_EXCEPTION;
                        unset($result[$k]);
```

The problem is `!Carrier::checkDeliveryPriceByPrice(...)` . This is the problem. This statement is using implicit casting and this is the issue we need to fix.

> As I tried to understand the goal of this method, I checked in our codebase how it is used, and the result of this method is always used as a boolean indeed 😓 So I understand where your fix comes from

I disagree with @matthieu-rolland : this function is a public function in Carrier class. Maybe five hundred module developers use it to retrieve the shipping price today. It's not because in the Core we use it as a boolean that everybody that uses this function use it as a boolean. Unless there is a very good reason to it, I don't see why we should modify the behavior of this code when it returns an integer different than zero.

File: classes/Carrier.php (hunk 1 above)
Comment (PierreRambaud) [reply classes/Carrier.php:330]: @Progi1984 It changes the method signature as it wasn't typed.
A signature change when to return type change and break previous behavior.
Type parameters and returns weren't working on PHP5.6, the code is old and wasn't designed with type.

If we do
```php
public static function checkDeliveryPriceByWeight($moduleName, $isStrong, $blablabla)
```

The signature doesn't change, but we don't expect the same parameter, so it's a BC.

**Comment by jf-viguier** [comment]
@matks ok, thanks.
This two public function in Carrier class don't return the same thing if a module is hooked or not:
- No module : return a boolean
- A module is hooked : return a shipping cost 

Do you really thing that this five hundred module developers want this ? Indeed, there is no way for this developers to know that a module is hooked or not.
I think that this two hooks are unusable right now in the explained context.

Do we really need to keep this buggy behavior ? Version 8 can have BC, may be it's a god time to fix this ?

**Comment by jf-viguier** [comment]
MAy I have another maintener point of view ? @Progi1984 ? I'll adapt the pr according to your decision.

**Comment by jf-viguier** [comment]
@matks Please check my last commit. I've removed implicit casting, this should fix the bug if maintainers decide to keep old beahavior and make the code cleaner anyway.

**Review by Progi1984 — DISMISSED** [review]

**Comment by Progi1984** [comment]
I approve this PR : the check is an assertion, and would have to return a `boolean` and not an `integer|boolean`.

**Comment by jf-viguier** [comment]
Thanks @Progi1984 you rocks

**Review by atomiix — DISMISSED** [review]

**Comment by jf-viguier** [comment]
Thanks @atomiix, ready for QA ?

File: classes/Carrier.php [hunk 2]
```diff
@@ -321,13 +321,13 @@ public function getDeliveryPriceByWeight($total_weight, $id_zone)
     }
 
     /**
-     * Get delivery price by total weight.
+     * Check delivery price by total weight.
```
Comment (eternoendless) [inline classes/Carrier.php]:
```suggestion
     * Check if the carrier is available for a given weight and zone
```

File: classes/Carrier.php [hunk 3]
```diff
@@ -426,14 +426,14 @@ public function getDeliveryPriceByPrice($order_total, $id_zone, $id_currency = n
     }
 
     /**
-     * Get delivery price for a given order.
+     * Check delivery price for a given order.
```
Comment (eternoendless) [inline classes/Carrier.php]:
```suggestion
     * Check if the carrier is available for a given order total, currency and zone
```

**Review by Progi1984 — APPROVED** [review]

**Review by eternoendless — APPROVED** [review]

**Comment by eternoendless** [comment]
I approve this because `checkDeliveryPriceBy**()` is clearly meant to return bool. `getDeliveryPriceBy**()` has been there for ages now, so the risk is acceptable.

**Comment by jf-viguier** [comment]
Thanks @eternoendless

**Review by PierreRambaud — APPROVED** [review]

**Review by HanaRebaiQA — COMMENTED** [review]
Hello @jf-viguier 

I have tested this PR. The issue is fixed.
Here the checked cases : 
1. Enable the `actionDeliveryPriceByWeight` hook and verify with 
-different carriers : OK
-multistore : OK
-free shipping enabled / disabled : OK
-enable the `actionDeliveryPriceByPrice` hook and disable the current one : OK

2. Enable the `actionDeliveryPriceByPrice` hook and verify with 
-different carriers : OK
-multistore : OK
-free shipping enabled / disabled : OK
-enable the `actionDeliveryPriceByWeight` hook and disable the current one : OK 

3. Both `actionDeliveryPriceByWeight` and `actionDeliveryPriceByPrice` are enabled : OK

image
image
https://watch.screencastify.com/v/ubbCj8yO7vYfKMlX2xF1
https://watch.screencastify.com/v/Xaf47fBO4el8zWK2XCIJ

So, QA :heavy_check_mark: 

Thanks!

**Review by HanaRebaiQA — APPROVED** [review]
Hello @jf-viguier 

I have tested this PR. The issue is fixed.
Here the checked cases : 
1. Enable the `actionDeliveryPriceByWeight` hook and verify with 
-different carriers : OK
-multistore : OK
-free shipping enabled / disabled : OK
-enable the `actionDeliveryPriceByPrice` hook and disable the current one : OK

2. Enable the `actionDeliveryPriceByPrice` hook and verify with 
-different carriers : OK
-multistore : OK
-free shipping enabled / disabled : OK
-enable the `actionDeliveryPriceByWeight` hook and disable the current one : OK 

3. Both `actionDeliveryPriceByWeight` and `actionDeliveryPriceByPrice` are enabled : OK

image
image
https://watch.screencastify.com/v/ubbCj8yO7vYfKMlX2xF1
https://watch.screencastify.com/v/Xaf47fBO4el8zWK2XCIJ

So, QA :heavy_check_mark: 

Thanks!

**Comment by jf-viguier** [comment]
Thanks @HanaRebaiQA

**Comment by Progi1984** [comment]
Thanks @jf-viguier & @HanaRebaiQA
